src/mlir/dialect/VeronaOps.td

```diff
@@ -5,36 +5,240 @@
+
+      The following example reads the value of field `f` from the reference at
+      `%x`.
+
+      ```mlir
+        %y = verona.field_read %x["f"] : !verona.C -> !verona.D
+      ```
+    }];
+
+    let arguments = (ins Verona_Type : $origin, StrAttr : $field);
+    let results = (outs Verona_Type : $output);
+
+    let assemblyFormat = [{$origin `[` $field `]` attr-dict `:` type($origin) `->` type($output)}];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_FieldWriteOp : Verona_Op<"field_write"> {
+    let summary = "Write to a field through a reference";
+    let description = [{
+      Write a new value to a field. The field's old value is returned.
+    }];
+
+    let arguments = (ins
+      Verona_Type : $origin,
+      Verona_Type : $value,
+      StrAttr : $field
+    );
+    let results = (outs Verona_Type : $output);
 
     let assemblyFormat = [{
-        $input attr-dict `:` type($input)
+        $origin `[` $field `]` `,` $value attr-dict `:` type($origin) `->` type($value) `->` type($output)
+    }];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_TidyOp : Verona_Op<"tidy"> {
+    let summary = "Perform garbage-collection on a region";
+    let description = [{
+      Perform garbage-collection in the specified region. The actual semantics
+      depend on the kind of region. For example, in an arena region, this could be
+      a no-op.
+    }];
+
+    let arguments = (ins Verona_Type : $region);
+    let assemblyFormat = [{$region attr-dict `:` type($region)}];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_DropOp : Verona_Op<"drop"> {
+    let summary = "Drop a value";
+    let description = [{
+      Drop a value, deallocating the referenced region if the value was an owned
+      reference.
+    }];
+
+    let arguments = (ins Verona_Type : $region);
+    let assemblyFormat = [{$region attr-dict `:` type($region)}];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_AllocateRegionOp : Verona_Op<"new_region"> {
+    let summary = "Allocate an object in a new region";
+    let description = [{
+      Allocate a new object of the given class in a new region.
+
+      The operation needs to be provided the list of field names, and for each
+      field, the SSA value to use to initialize it.
+
+      The returned value is an owned reference to the allocated object, which is
+      also the entrypoint to the region.
+
+      The following example allocates an object of class C in a new region. The
+      fields f and g of the object are initialized respectively with the values
+      %x and %y.
+
+      ```mlir
+        %w = verona.new_region @C["f", "g"](%x, %y : !verona.A, !verona.B) : !verona.C
+      ```
+
+      TODO: the syntax is a little ugly, something that looks like a record
+      eg. `verona.new_region @C[ "f" = %x : !verona.A, "g" = %y : !verona.B ]`,
+      binding field names to SSA values might be nicer. Probably needs a custom
+      parser though.
```

Comment:
MLIR's default syntax is not super consistent. We'll have to use a custom parser at some point, but it should be fine now for the first pass.